### ISSUE TITLE
Models performance decreases with releases?

### ISSUE THREAD

**Issue by Louis-Mozart** [dice-group/dice-embeddings#429]
Two years ago (dicee version 0.1.4), when running the following command for dice embeddings

` python main.py --path_dataset_folder KGs/UMLS --model DeCaL --auto_signature --num_epochs 250 --embedding_dim 16 --p 6 --q 0 --r 1 --save_model_at_every_epoch 999
`
We could get the following results

```
Epoch 249: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  3.18it/s, loss=0.0375]
Training Runtime: 1.389 minutes.

*** Save Trained Model ***
Took 0.0030 secs | Current Memory Usage  883.49 in MB
Total Runtime: 83.488 seconds
Evaluate DeCaL on Train set: Evaluate DeCaL on Train set
{'H@1': 0.8975268404907976, 'H@3': 0.9857170245398773, 'H@10': 0.9959739263803681, 'MRR': np.float64(0.9420816344870241)}
Evaluate DeCaL on Validation set: Evaluate DeCaL on Validation set
{'H@1': 0.8182515337423313, 'H@3': 0.9470858895705522, 'H@10': 0.9785276073619632, 'MRR': np.float64(0.8865539462015206)}
Evaluate DeCaL on Test set: Evaluate DeCaL on Test set
{'H@1': 0.8018154311649016, 'H@3': 0.9523449319213313, 'H@10': 0.9848714069591528, 'MRR': np.float64(0.8798998702040874)}
Total Runtime: 85.634 seconds
```

With the current version, running the same command 

`python dicee/scripts/run.py --dataset_dir KGs/UMLS --trainer PL --scoring_technique KvsAll --model DeCaL --auto_signature --embedding_dim 16  --num_epochs 250 --p 6 --q 0 --r 1 --random_seed 0 --batch_size 1024 --optim Adam --num_core 2 --save_model_at_every_epoch 999`

Gives 
```

Epoch 249: 100%|██████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 12.15it/s, loss_step=0.0346, loss_epoch=0.0326]`Trainer.fit` stopped: `max_epochs=250` reached.
Epoch 249: 100%|██████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 11.32it/s, loss_step=0.0346, loss_epoch=0.0326]
2026-08-26 14:47:08,581 INFO dicee.callbacks: Training Runtime: 36.525 seconds.
2026-08-26 14:47:08,582 INFO dicee.executer: *** Save Trained Model ***
2026-08-26 14:47:08,583 INFO dicee.static_funcs: Took 0.0007 secs | Current Memory Usage 524.53786 MB
2026-08-26 14:47:08,583 INFO dicee.executer: Total Runtime: 36.722 seconds
2026-08-26 14:47:08,635 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Train set:
2026-08-26 14:47:09,192 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Train set
2026-08-26 14:47:09,192 INFO dicee.evaluation.evaluator: {'H@1': 0.8913918711656442, 'H@3': 0.9899348159509203, 'H@10': 0.9981786809815951, 'MRR': 0.9412327185885137}
2026-08-26 14:47:09,192 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Validation set:
2026-08-26 14:47:09,315 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Validation set
2026-08-26 14:47:09,315 INFO dicee.evaluation.evaluator: {'H@1': 0.7507668711656442, 'H@3': 0.9256134969325154, 'H@10': 0.9739263803680982, 'MRR': 0.8459464629765707}
2026-08-26 14:47:09,315 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Test set:
2026-08-26 14:47:09,492 INFO dicee.evaluation.evaluator: Evaluate DeCaL on Test set
2026-08-26 14:47:09,492 INFO dicee.evaluation.evaluator: {'H@1': 0.7443267776096822, 'H@3': 0.9311649016641452, 'H@10': 0.9818456883509834, 'MRR': 0.8424539202599624}
2026-08-26 14:47:09,493 INFO dicee.executer: Total Runtime: 37.631 seconds
```
i.e., we can see a nice improve on the runtime but the performance, especially on test and validation, dropped significantly. 

This observation is also visible to other models e.g. See below the resuts of ComplEx on the KINSHIP datasets

On dicee (0.1.4)
```
Epoch 249: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 25.01it/s, loss=0.0745]
Training Runtime: 33.511 seconds.

*** Save Trained Model ***
Took 0.0054 secs | Current Memory Usage  883.17 in MB
Total Runtime: 33.654 seconds
Evaluate ComplEx on Train set: Evaluate ComplEx on Train set
{'H@1': 0.6798923220973783, 'H@3': 0.8689723782771536, 'H@10': 0.9734901685393258, 'MRR': np.float64(0.7862436616772492)}
Evaluate ComplEx on Validation set: Evaluate ComplEx on Validation set
{'H@1': 0.6118913857677902, 'H@3': 0.8314606741573034, 'H@10': 0.9616104868913857, 'MRR': np.float64(0.7375810874014029)}
Evaluate ComplEx on Test set: Evaluate ComplEx on Test set
{'H@1': 0.5940409683426443, 'H@3': 0.8212290502793296, 'H@10': 0.962756052141527, 'MRR': np.float64(0.725130845196416)}
Total Runtime: 34.611 seconds
```

On the current version 

```
Epoch 249: 100%|██████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 26.66it/s, loss_step=0.0793, loss_epoch=0.0758]`Trainer.fit` stopped: `max_epochs=250` reached.
Epoch 249: 100%|██████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 25.23it/s, loss_step=0.0793, loss_epoch=0.0758]
2026-08-26 15:00:10,710 INFO dicee.callbacks: Training Runtime: 29.017 seconds.
2026-08-26 15:00:10,714 INFO dicee.executer: *** Save Trained Model ***
2026-08-26 15:00:10,715 INFO dicee.static_funcs: Took 0.0009 secs | Current Memory Usage 525.73389 MB
2026-08-26 15:00:10,715 INFO dicee.executer: Total Runtime: 29.245 seconds
2026-08-26 15:00:10,793 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Train set:
2026-08-26 15:00:11,644 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Train set
2026-08-26 15:00:11,644 INFO dicee.evaluation.evaluator: {'H@1': 0.6592345505617978, 'H@3': 0.8579119850187266, 'H@10': 0.9712078651685393, 'MRR': 0.7709250859920975}
2026-08-26 15:00:11,645 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Validation set:
2026-08-26 15:00:11,878 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Validation set
2026-08-26 15:00:11,878 INFO dicee.evaluation.evaluator: {'H@1': 0.5749063670411985, 'H@3': 0.8033707865168539, 'H@10': 0.9531835205992509, 'MRR': 0.7062853169870879}
2026-08-26 15:00:11,878 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Test set:
2026-08-26 15:00:12,022 INFO dicee.evaluation.evaluator: Evaluate ComplEx on Test set
2026-08-26 15:00:12,022 INFO dicee.evaluation.evaluator: {'H@1': 0.5670391061452514, 'H@3': 0.8086592178770949, 'H@10': 0.9622905027932961, 'MRR': 0.7053928904811408}
2026-08-26 15:00:12,022 INFO dicee.executer: Total Runtime: 30.551 seconds
```

Below, I have attached the configuration file of the ComplEx model for both versions.

[configuration_dicee_0_1_4_Complex.json](https://github.com/user-attachments/files/31472429/configuration_dicee_0_1_4_Complex.json)

[configuration_dicee_current_Complex.json](https://github.com/user-attachments/files/31472481/configuration_dicee_current_Complex.json)

**Comment by Demirrr** [comment]
Thank you for the issue. I am on it

**Comment by Demirrr** [comment]
I bisected this using a controlled protocol (identical hyperparameters across versions, only the code commit varies) against `ComplEx` on `KGs/KINSHIP` (Test H@1), with `git bisect run` driving 250-epoch training runs automatically.

**Two commits, both from 2025-12-15 (same author, back-to-back), fully explain the regression:**

| Stage | Test H@1 |
|---|---|
| `v1.0.1.4` (`dicee==0.1.4`) baseline | 0.6020 |
| after `631f3e39` "Fix order-dependent optimization issue (#356)" | 0.5843 |
| after `ecdba8c0` "Make vocabulary construction order-independent by sorting entities and relations alphabetically" | **0.5670** (exact match to current `develop`) |

That's a complete decomposition — these two commits account for the entire measured gap across the ~300 commits bisected; nothing else in between mattered.

**Root cause — not a bug, a side effect of a legitimate fix:** both commits replace incidental, I/O-order-dependent construction (entity/relation index assignment, and the `KvsAll` `(head, relation) → tails` dict) with a canonical **alphabetically-sorted** order. That's a real correctness improvement — training used to silently depend on the order triples appeared in the input file. But it means a *fixed* `random_seed` no longer reproduces the same trained model as before: which integer index (and thus which embedding row / random init) each entity gets is now different, and so is which pairs get grouped into which mini-batch. For `random_seed=0` on these small benchmarks, the new deterministic mapping happens to land in a somewhat worse-generalizing basin than the old one did — train fit stays similar, but validation/test suffer, exactly the pattern reported here.

**DeCaL/UMLS half of this issue is already partially fixed:** `#428` (merged 2026-08-18, "Fix DeCaL initialization") recovers most of that gap. At current `develop`, DeCaL/UMLS Val H@1 is 0.794, vs. 0.751 in the pre-`#428` buggy state, vs. 0.818 in `v1.0.1.4`. The remaining ~0.02 residual there is consistent with the same seed-remapping effect described above.

**Suggestion:** the sorting fix (`631f3e39`/`ecdba8c0`) is correct and worth keeping — I wouldn't revert it. The real issue is that "same `random_seed`" no longer guarantees "same result across versions," so it's worth a quick multi-seed sanity run to confirm this is seed-lottery noise rather than a systematic bias, and re-baselining `tests/test_regression_*.py`'s bounds, which were tuned against the old order-dependent index/batch assignment.
